scale_build/checkout.py

```diff
@@ -0,0 +1,43 @@
+import logging
+
+from .config import BRANCH_OVERRIDES, TRY_BRANCH_OVERRIDE
+from .exceptions import CallError
+from .utils.git_utils import branch_exists_in_repository, retrieve_git_remote_and_sha, update_git_manifest
+from .utils.package import get_packages
+
+
+logger = logging.getLogger(__name__)
+
+
+def checkout_sources():
+    info = retrieve_git_remote_and_sha('.')
+    update_git_manifest(info['url'], info['sha'], 'w')
+    logger.info('Starting checkout of source')
+
+    for package in get_packages():
+        gh_override = BRANCH_OVERRIDES.get(package.name)
+
+        # TRY_BRANCH_OVERRIDE is a special use-case. It allows setting a branch name to be used
+        # during the checkout phase, only if it exists on the remote.
+        #
+        # This is useful for PR builds and testing where you want to use defaults for most repos
+        # but need to test building of a series of repos with the same experimental branch
+        #
+        if TRY_BRANCH_OVERRIDE:
+            retries = 3
+            while retries:
+                try:
+                    if branch_exists_in_repository(package.origin, TRY_BRANCH_OVERRIDE):
+                        gh_override = TRY_BRANCH_OVERRIDE
+                except CallError:
+                    retries -= 1
+                    logger.debug(
+                        'Failed to determine if %r branch exists for %r. Trying again',
+                        TRY_BRANCH_OVERRIDE, package.origin
+                    )
+                    if not retries:
+                        logger.debug('Unable to determine if %r branch exists in 3 attempts.')
```

Comment:
`%r` is not used

Why can't we just fail here? If I specified `TRY_BRANCH_OVERRIDE`, I want my build to use `TRY_BRANCH_OVERRIDE` or to fail if it does not exist. I don't want it to fake a build by giving me an image that uses some other branch. Will I have to read logs every time to see if it gave me a build I ordered?

Reply:
My intention was not to change existing behavior. The builder relies on this behavior right now for generating builds. Can you please let me know how you think we should perhaps cater to this use case ? Otherwise the user would have to specify an override for each repo ( we have ~44 almost )

Reply:
> Otherwise the user would have to specify an override for each repo ( we have ~44 almost )

Why?

Reply:
I am not sure why this was added but i think the use case was to see if any of the sources have `TRY_BRANCH_OVERRIDE` branch and if they do use it so that we have a build which covers the ticket in question. Like say the value set is NAS-11111, now if the ticket has changes to middleware and some other repos, just specifying this in the builder PR, we can expect the resulting build to have all the changes from all these repos. @kmoore134 can you please confirm ?

Reply:
That is used by CI. SO that if you issue a pull-request to truenas/scale-build for branch NAS-111111, and there is an identical branch on truenas/middleware called NAS-111111, it'll use same branches on both for that build. Thats how to handle complex PR's that rely on other WiP branches or PRs to get test runs.